c++/src/kj/compat/http-test.c++

```diff
@@ -4694,5 +4705,295 @@ KJ_TEST("HttpServer.listenHttp() doesn't prematurely terminate if an accepted co
+    auto io = tunnel.accept(statusCodeToSend, "OK", HttpHeaders(headerTable));
+    return io->pumpTo(*io).ignoreResult().attach(kj::mv(io));
+  }
+
+private:
+  HttpHeaderTable& headerTable;
+  uint statusCodeToSend;
+};
+
+class ConnectRejectService final: public HttpService {
+public:
+  ConnectRejectService(HttpHeaderTable& headerTable, uint statusCodeToSend = 400)
+      : headerTable(headerTable),
+        statusCodeToSend(statusCodeToSend) {
+    KJ_ASSERT(statusCodeToSend >= 300);
+  }
+
+  uint connectCount = 0;
+
+  kj::Promise<void> request(
+      HttpMethod method, kj::StringPtr url, const HttpHeaders& headers,
+      kj::AsyncInputStream& requestBody, Response& response) override {
+    KJ_UNIMPLEMENTED("Regular HTTP requests are not implemented here.");
+  }
+
+  kj::Promise<void> connect(kj::StringPtr host, const HttpHeaders& headers,
+                           ConnectResponse& tunnel) override {
+    // A simple CONNECT echo. It will always accept, and whatever data it
+    // receives will be echoed back.
+    connectCount++;
+    auto out = tunnel.reject(statusCodeToSend, "Failed"_kj, HttpHeaders(headerTable));
+    return out->write("boom", 4).attach(kj::mv(out));
+  }
+
+private:
+  HttpHeaderTable& headerTable;
+  uint statusCodeToSend;
+};
+
+KJ_TEST("Simple CONNECT Server works") {
+  KJ_HTTP_TEST_SETUP_IO;
+
+  auto pipe = KJ_HTTP_TEST_CREATE_2PIPE;
+
+  kj::TimerImpl timer(kj::origin<kj::TimePoint>());
+
+  HttpHeaderTable table;
+  ConnectEchoService service(table);
+  HttpServer server(timer, table, service);
+
+  auto listenTask = server.listenHttp(kj::mv(pipe.ends[0]));
+
+  auto msg = "CONNECT https://example.org HTTP/1.1\r\n"
+             "\r\n"
+             "hello"_kj;
+
+  pipe.ends[1]->write(msg.begin(), msg.size()).wait(waitScope);
+  pipe.ends[1]->shutdownWrite();
+
+  expectRead(*pipe.ends[1],
+             "HTTP/1.1 200 OK\r\n"
+             "\r\n"
+             "hello"_kj).wait(waitScope);
+
+  expectEnd(*pipe.ends[1]);
+
+  listenTask.wait(waitScope);
+
+  KJ_ASSERT(service.connectCount == 1);
+}
+
+KJ_TEST(" ") {
+  KJ_HTTP_TEST_SETUP_IO;
+
+  auto pipe = KJ_HTTP_TEST_CREATE_2PIPE;
+
+  kj::TimerImpl timer(kj::origin<kj::TimePoint>());
+
+  HttpHeaderTable table;
+  ConnectEchoService service(table);
+  HttpServer server(timer, table, service);
+
+  auto listenTask = server.listenHttp(kj::mv(pipe.ends[0]));
+
+  auto client = newHttpClient(table, *pipe.ends[1]);
+
+  HttpHeaderTable clientHeaders;
+  // Initiates a CONNECT with the echo server. Once established, sends a bit of data
+  // and waits for it to be echoed back.
+  client->connect("https://example.org"_kj, HttpHeaders(clientHeaders))
+      .then([&](HttpClient::ConnectResponse response) {
+    KJ_ASSERT(response.statusCode == 200);
+    KJ_ASSERT(response.statusText == "OK"_kj);
+
+    auto& io = KJ_ASSERT_NONNULL(response.connectionOrBody.tryGet<kj::Own<kj::AsyncIoStream>>());
+    io->write("hello", 5).poll(waitScope);
```

Comment:
Why `.poll()` rather than `.wait()`? This will cancel the promise if it doesn't complete immediately, which seems weird.

Reply:
ah right, I mean to leave a comment about this but forgot. Using wait() here is causing an error to be reported saying there's nothing to wait on. Wasn't quite sure what was going on here yet.